src/imp_pl.rs

```diff
@@ -121,18 +112,47 @@ impl<T> OnceCell<T> {
     }
 }
 
+struct Guard<'a> {
+    state: &'a AtomicU8,
+    new_state: u8,
+}
+
+impl<'a> Drop for Guard<'a> {
+    fn drop(&mut self) {
+        self.state.store(self.new_state, Ordering::Release);
+    }
+}
+
 // Note: this is intentionally monomorphic
 #[inline(never)]
-fn initialize_inner(
-    mutex: &Mutex<()>,
-    is_initialized: &AtomicBool,
-    init: &mut dyn FnMut() -> bool,
-) {
-    let _guard = mutex.lock();
-
-    if !is_initialized.load(Ordering::Acquire) {
-        if init() {
-            is_initialized.store(true, Ordering::Release);
+fn initialize_inner(state: &AtomicU8, init: &mut dyn FnMut() -> bool) {
+    let key = state as *const _ as usize;
+    loop {
+        let exchange =
+            state.compare_exchange(INCOMPLETE, RUNNING, Ordering::Acquire, Ordering::Acquire);
```

Comment:
```suggestion
            state.compare_exchange(INCOMPLETE, RUNNING, Ordering::Acquire, Ordering::Relaxed);
```

You could also consider `compare_exchange_weak`, though I don't know if there's any benefit in doing so since you will need to `continue` on `Err(INCOMPLETE)` anyway.

Reply:
Hm, this I think also needs to be acquire, otherwise `Err(COMPLETE) => return,` branch won't synchronize-with the corresponding Release store, and subsequent `get_unchecked` would read garbage. Both cases are great to check if loom works though!